services/proxy/pkg/userroles/oidcroles.go

```diff
@@ -29,6 +31,58 @@ func NewOIDCRoleAssigner(opts ...Option) UserRoleAssigner {
 	}
 }
 
+func extractRoles(rolesClaim string, claims map[string]interface{}) (map[string]struct{}, error) {
+
+	claimRoles := map[string]struct{}{}
+	// happy path
+	value, _ := claims[rolesClaim].(string)
+	if value != "" {
+		claimRoles[value] = struct{}{}
+		return claimRoles, nil
+	}
+
+	// try splitting path at .
+	segments := oidc.SplitWithEscaping(rolesClaim, ".", "\\")
+
+	subclaims := claims
+	lastSegment := len(segments) - 1
+	for i := range segments {
+		if i < lastSegment {
+			if castedClaims, ok := subclaims[segments[i]].(map[string]interface{}); ok {
+				subclaims = castedClaims
+			} else if castedClaims, ok := subclaims[segments[i]].(map[interface{}]interface{}); ok {
+				subclaims = make(map[string]interface{}, len(castedClaims))
+				for k, v := range castedClaims {
+					if s, ok := k.(string); ok {
+						subclaims[s] = v
+					} else {
+						return nil, fmt.Errorf("could not walk claims path, key '%v' is not a string", k)
+					}
+				}
```

Comment:
Isn't this the same code as in `oidc` package? Can't we just extract this to a function called here?

Reply:
slightly different. The oidc package tries to get a single value string. this cods tries to get a map of roles. I extracted and reused `SplitWithEscaping`. If you see a good way to make that code more reusable go for it.